clients/azuredevopsrepo/zip.go

```diff
@@ -0,0 +1,251 @@
+			z.errSetup = sce.WithMessage(sce.ErrScorecardInternal, err.Error())
+			return
+		}
+	})
+
+	return z.errSetup
+}
+
+func (z *zipHandler) getZipfile() error {
+	tempDir, err := os.MkdirTemp("", repoDir)
+	if err != nil {
+		return fmt.Errorf("os.MkdirTemp: %w", err)
+	}
+	repoFile, err := os.CreateTemp(tempDir, repoFilename)
+	if err != nil {
+		return fmt.Errorf("%w io.Copy: %w", errZipNotFound, err)
+	}
+	defer repoFile.Close()
+
+	// The zip download API is not exposed in the Azure DevOps Go SDK, so we need to construct the request manually.
+	baseURL := fmt.Sprintf(
+		"https://%s/%s/%s/_apis/git/repositories/%s/items",
+		z.repourl.host,
+		z.repourl.organization,
+		z.repourl.project,
+		z.repourl.id)
+
+	queryParams := url.Values{}
+	queryParams.Add("path", "/")
+	queryParams.Add("download", "true")
+	queryParams.Add("api-version", "7.1-preview.1")
+	queryParams.Add("resolveLfs", "true")
+	queryParams.Add("$format", "zip")
+
+	if z.repourl.commitSHA == "HEAD" {
+		queryParams.Add("versionDescriptor.versionType", "branch")
+		queryParams.Add("versionDescriptor.version", z.repourl.defaultBranch)
+	} else {
+		queryParams.Add("versionDescriptor.versionType", "commit")
+		queryParams.Add("versionDescriptor.version", z.repourl.commitSHA)
+	}
+
+	parsedURL, err := url.Parse(baseURL + "?" + queryParams.Encode())
+	if err != nil {
+		return fmt.Errorf("url.Parse: %w", err)
+	}
+
+	req := &http.Request{
+		Method: http.MethodGet,
+		URL:    parsedURL,
+	}
+	res, err := z.client.SendRequest(req)
+	if err != nil {
+		return fmt.Errorf("client.SendRequest: %w", err)
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("%w: status code %d", errUnexpectedStatusCode, res.StatusCode)
+	}
+
+	if _, err := io.Copy(repoFile, res.Body); err != nil {
+		return fmt.Errorf("io.Copy: %w", err)
+	}
+
+	z.tempDir = tempDir
+	z.tempZipFile = repoFile.Name()
+
+	return nil
+}
+
+func (z *zipHandler) getLocalPath() (string, error) {
+	if err := z.setup(); err != nil {
+		return "", fmt.Errorf("error during zipHandler.setup: %w", err)
+	}
+	absTempDir, err := filepath.Abs(z.tempDir)
+	if err != nil {
+		return "", fmt.Errorf("error during filepath.Abs: %w", err)
+	}
+	return absTempDir, nil
+}
+
+func (z *zipHandler) extractZip() error {
+	in, err := os.OpenFile(z.tempZipFile, os.O_RDONLY, 0o644)
+	if err != nil {
+		return fmt.Errorf("os.OpenFile: %w", err)
+	}
+
+	stat, err := in.Stat()
+	if err != nil {
+		return fmt.Errorf("os.File.Stat: %w", err)
+	}
+
+	zipReader, err := zip.NewReader(in, stat.Size())
+	if err != nil {
+		return fmt.Errorf("zip.NewReader: %w", err)
+	}
+
+	z.files = make([]string, 0, len(zipReader.File))
+	for _, file := range zipReader.File {
+		//nolint:gosec // G305: Handling of file paths is done below
+		filenamepath := filepath.Join(z.tempDir, file.Name)
+		if !strings.HasPrefix(filepath.Clean(filenamepath), filepath.Clean(z.tempDir)) {
+			return errInvalidFilePath
+		}
```

Comment:
No action required, just posting for discussion / posterity.

Go takes a slightly different sanitization technique, but only with a godebug value set:
https://go.dev/doc/godebug#go-120
> Go 1.20 introduced support for rejecting insecure paths in tar and zip archives, controlled by the [tarinsecurepath setting](https://go.dev/pkg/archive/tar/#Reader.Next) and the [zipinsecurepath setting](https://go.dev/pkg/archive/zip/#NewReader). These default to tarinsecurepath=1 and zipinsecurepath=1, preserving the behavior of earlier versions of Go. A future version of Go may change the defaults to tarinsecurepath=0 and zipinsecurepath=0.